### PR TITLE
Implement "Create" Instrument Listing Feature

### PR DESCRIPTION
**Pull request by moelgazzar**
This feature creates new instrument listings, which can be accessed at **/instruments/new**

- Added "new" and "create" actions in InstrumentsController.
- Created "new.html.erb" view for the instrument form "using simple_form".
- Updated the Instrument model with validations for required fields (name, address, category, price, description).
- Modified the routes to include routes for "new" and "create" actions.

### CODE REVIEW

**Comment by p-guelfi** [comment]
Looks good!

File: app/controllers/instruments_controller.rb [hunk 1]
```diff
@@ -1,9 +1,35 @@
 class InstrumentsController < ApplicationController
+  # This ensures that users are logged in for actions that modify data
+  before_action :authenticate_user!, only: [:new, :create]
```
Comment (DavidDicke) [inline app/controllers/instruments_controller.rb]: I think in the lecture they taught that the best practice would be the white listing approach. So protect every route in the applications controller as a default and then we would not have to insert it here, but do `skip_before_action :authenticate_user!, only: %i[index home]`

File: app/controllers/instruments_controller.rb [hunk 2]
```diff
@@ -1,9 +1,35 @@
...
+  end
+
+  def create
+    @instrument = Instrument.new(instrument_params)
+    @instrument.user = current_user
+
+    if @instrument.save
+      redirect_to instrument_path(@instrument), notice: 'Instrument successfully created!'
+    else
+      render :new, alert: 'There was an issue creating your instrument listing.'
```
Comment (DavidDicke) [inline app/controllers/instruments_controller.rb]: We should add the `status: :unprocessable_entity`

File: app/models/instrument.rb [hunk 3]
```diff
@@ -1,3 +1,9 @@
 class Instrument < ApplicationRecord
   belongs_to :user
+
```
Comment (DavidDicke) [inline app/models/instrument.rb:4]: we can add the `has_many_attached :photos` here as well.

File: app/models/instrument.rb [hunk 4]
```diff
@@ -1,3 +1,9 @@
...
+  validates :name, presence: true
+  validates :description, presence: true
+  validates :category, presence: true
+  validates :address, presence: true
+  validates :price, numericality: { greater_than_or_equal_to: 0 }
```
Comment (DavidDicke) [inline app/models/instrument.rb]: also: `presence: true`

File: app/views/instruments/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<%= simple_form_for @instrument do |f| %>
+  <%= f.input :name, label: "Instrument Name" %>
+  <%= f.input :description, label: "Description" %>
+  <%= f.input :address, label: "Address" %>
+  <%= f.input :price, label: "Price" %>
+  <%= f.input :category, label: "Category" %>
+
```
Comment (DavidDicke) [inline app/views/instruments/new.html.erb:8]: If we have the attached photos in model, we should have an upload field here: `<%= f.input :photos, as: :file, input_html: { multiple: true } %>`

File: app/views/instruments/show.html.erb [hunk 6]
```diff
@@ -3,6 +3,8 @@
```
Comment (DavidDicke) [inline app/views/instruments/show.html.erb:3]: Here we would iterate over the photos and show them with the cl_image_tag maybe as a carousel.

**Review by DavidDicke — COMMENTED** [review]
Hey I added some comments regarding best practices and the Photos that we want to add to instruments.

**Comment by moelgazzar** [comment]
Hi @DavidDicke, @p-guelfi,

I have addressed all the feedback and made the following adjustments:

- Centralized authentication by adding `before_action :authenticate_user!` in `ApplicationController`.
- Whitelisted the `index` and `show` actions in `InstrumentsController` using `skip_before_action :authenticate_user!`.
- Updated the `create` action in `InstrumentsController`:
  - Added `flash[:alert]` to show an error message when form submission fails.
  - Used `status: :unprocessable_entity` to give the correct HTTP status for invalid form submissions.
- Updated the `new.html.erb` view to display `flash[:alert]` messages when an issue occurs during instrument creation.

Let me know if there’s anything else to adjust.

Thanks!

**Pull request merged**